File: Python/downloadFiles.py
'''
DISCLAIMER OF WARRANTIES.
 This code is sample code created by IBM Corporation. IBM grants you a
 nonexclusive copyright license to use this sample code example. This
 sample code is not part of any standard IBM product and is provided to you
 solely for the purpose of assisting you in the development of your
 applications. This example has not been thoroughly tested under all
 conditions. IBM, therefore cannot guarantee nor may you imply reliability,
 serviceability, or function of these programs. The code is provided "AS IS",
 without warranty of any kind. IBM shall not be liable for any damages
 arising out of your or any other parties use of the sample code, even if IBM
 has been advised of the possibility of such damages. If you do not agree with
 these terms, do not use the sample code.

 Copyright IBM Corp. 2019 All Rights Reserved.

 To run, see README.md
'''
import json, os
import requests
import datetime as dt
from loggingHandler import logger
from readConfigJSON import readJSON
import time, copy
from ssl import SSLError
from base64 import b64encode

'''
    Function to check file status if completed, failed or inprogress
'''
def checkStatus(configuration_settings, analyzerId):
    # Make request
    headers = {
        'apiKey': configuration_settings['api_key']
    }

    if "function_id" in configuration_settings:
            credentials = ('%s:%s' % (
                configuration_settings["function_id"], configuration_settings["password"]))
            encoded_credentials = b64encode(credentials.encode('ascii'))
            headers['Authorization'] = 'Basic %s' % encoded_credentials.decode("utf-8")

    get_url = configuration_settings['main_url'] + "/{0}".format(analyzerId)
    try:
        response = requests.request("GET", get_url, headers=headers, verify=configuration_settings['ssl_verification'])
        if response.status_code >= 400:
            logger.error("An error occurred while trying to get file status.")
            return False, response.text
        else:
            response_json = json.loads(response.text)
            return True, response_json
    except SSLError as sslerror:
        logger.error("SSL error was thrown due to certificate failure, set ssl_verification to false in configuration config.json file.")
        logger.debug(sslerror, exc_info=True)
    except Exception as ex:
        print(ex)
        logger.error("An error occured when trying to get file")
        logger.debug(ex, exc_info=True)
        pass
    return False, {}

'''
    Function to download file and save in user's output_directory_path, based on the output type (pdf, utf8,json)
'''
def downloadFile(configuration_settings, analyzerId, output, output_path, filename):
    # Make request
    headers = {
        'apiKey': configuration_settings['api_key']
    }

    if "function_id" in configuration_settings:
        credentials = ('%s:%s' % (
            configuration_settings["function_id"], configuration_settings["password"]))
        encoded_credentials = b64encode(credentials.encode('ascii'))
        headers['Authorization'] = 'Basic %s' % encoded_credentials.decode("utf-8")

    get_url = configuration_settings['main_url'] + "/{0}/{1}".format(analyzerId, output.lower())
    try:
        response = requests.request("GET", get_url, headers=headers, verify=configuration_settings['ssl_verification'])
        if response.status_code >= 400:
            logger.error("An error occurred while trying to get file status.")
            return False, response.text
        else:
            file_name = copy.copy(filename)
            filename_output = "{0}.{1}".format(file_name, output.lower()) if output.lower() != "utf8" else "{0}.txt".format(file_name)
            filename_output = filename_output if output.lower() != "pdf" else "New_{0}.{1}".format(file_name, output.lower())
            file_output_path = os.path.join(configuration_settings["output_directory_path"], output_path, output.lower())
            filename_output_path = os.path.join(file_output_path, filename_output)
            if( not os.path.exists(file_output_path)):
                os.makedirs(file_output_path)
                logger.info("Created new output directory, "+ file_output_path)
            if(output.lower() == "json"):
                response_output = json.loads(response.text)
                if "data" in response_output:
                    json.dump(response_output["data"], open(filename_output_path, "w"), indent=4)
            elif output.lower() == "pdf":
                output_write = open(filename_output_path, "wb")
                output_write.write(response.content)
                output_write.close()
            else:
                output_write = open(filename_output_path, "w")
                try:
                    output_write.write(response.text)
                except:
                    output_write.write(response.text.encode('utf-8'))
                output_write.close()
            return True, ""
    except SSLError as sslerror:
        logger.error("SSL error was thrown due to certificate failure, set ssl_verification to false in configuration config.json file.")
        logger.debug(sslerror, exc_info=True)
    except Exception as ex:
        logger.error("An error occurred when trying to get file")
        logger.debug(ex, exc_info=True)
        pass
    return False, ""

'''
    Function to confirm if download outputs already exists
'''
def checkCompleted(latest_output_outputs, result):
    output_exists = []
    for output in latest_output_outputs:
        if (output in result):
            output_exists.append(output)
    if(len(output_exists) >= len(latest_output_outputs)):
        return True
    else:
        return False


'''
    Function to begin looking at the output.json file, for each content, check file processing status if ready, download the file, if not ready keep looping, till you get all the file, loop 30000 times and stop if not complete, so system resources are freed while looping, user's can always run downloadFiles.py again to continue downloading
'''
def downloadFiles():
    pending_completion = True
    configuration, configuration_settings = readJSON()
    if (configuration):
        loop = 0
        failed_download = []
        completed_download = []
        completed_count = 0
        output_json_path = os.path.join(os.getcwd(), "output.json")
        if(os.path.exists(output_json_path)):
            while pending_completion and loop < 1000:
                output_json = json.load(open(output_json_path, "rb"))
                loop += 1
                logger.info("Loop " + str(loop))
                if("output_results" in output_json and len(output_json["output_results"]) > 0):
                    output_json_result = output_json["output_results"]
                    new_output_json_result = []
                    for outresult in output_json_result:
                        result = outresult
                        try:
                            if "response" in result and "data" in json.loads(result["response"]) and "analyzerId" in json.loads(result["response"])["data"]:
                                if "download_success" not in result:
                                    response = json.loads(result["response"])
                                    path = result["path"]
                                    filename = result["filename"]
                                    analyzerId = response["data"]["analyzerId"]

                                    output_outputs = result["output_type"]
                                    latest_output_outputs = [output.replace("\"", "").upper() for output in
                                                             output_outputs]
                                    if("download_completed" not in result):
                                        completed = checkCompleted(latest_output_outputs, result)
                                        if(completed):
                                            result["download_completed"] = True
                                        else:
                                            logger.info("Checking status of analyzerId: " + analyzerId)
                                            status, result_response = checkStatus(configuration_settings, analyzerId)
                                            if (status):
                                                if ("data" in result_response and "statusDetails" in result_response[
                                                    "data"]):
                                                    status_result_response = result_response["data"]["statusDetails"]

                                                    done_output = []
                                                    for output in status_result_response:
                                                        if (output["type"] in latest_output_outputs and output[
                                                            "status"] == "Completed" and output["type"] not in result):
                                                            logger.info("Downloading {0} of analyzerId: {1}".format(output["type"], analyzerId))
                                                            response, reason = downloadFile(configuration_settings, analyzerId,
                                                                                    output["type"], path,
                                                                                    filename.rsplit(".")[0])

                                                            result[output["type"]] = response
                                                            if (not response):
                                                                result[output["type"] + "_error"] = reason
                                                            done_output.append(output["type"])
                                                        elif (output["type"] in latest_output_outputs and output[
                                                            "status"] == "Failed" and output["type"] not in result):

                                                            result[output["type"]] = False
                                                            result[output["type"] + "_error"] = output["status"]
                                                            done_output.append(output["type"])
                                                        elif output["type"] in result:
                                                            done_output.append(output["type"])
                                                    if (len(done_output) == len(latest_output_outputs)):
                                                        completed_download.append(True)

                                                    completed = checkCompleted(latest_output_outputs, result)
                                                    if(completed):
                                                        result["download_completed"] = True

                                            else:
                                                result["download_success"] = False
                                                result["download_failure_reason"] = result_response
                                                failed_download.append(True)
                            else:
                                logger.error("We could not find any information to download files from.")
                                result["download_success"] = False
                                result["download_failure_reason"] = "No available data to download from"
                                failed_download.append(True)
                        except:
                            result["download_success"] = False
                            result["download_failure_reason"] = "No analyzerID available to download results"
                            logger.error("No analyzerID available to download results. The file upload may have failed. File name: {0}".format(result["filename"]))
                            failed_download.append(True)

                        # new_output_json_result.append(result)
                        # print(len(new_output_json_result))
                        endtime = dt.datetime.now()
                        output_json["output_results"] = output_json_result
                        output_json["endtime"] = str(endtime)
                        json.dump(output_json, open("output.json", 'w'), indent=4)
                        completed_count = len(failed_download) + len(completed_download)

                    if (completed_count >= len(output_json_result)):
                        pending_completion = False
                    else:
                        if (loop >= 1000):
                            pending_completion = False
                            logger.error("Reached maximum number of download retries.")
                        else:
                            time.sleep(5)
                else:
                    pending_completion = False
                    logger.error("No results available to download.")
                    return True

            else:
                logger.info("Done downloading all output files to your output_directory_path")
                logger.info("Download status reported in output.json")
                if(loop < 999):
                    return True
                else:
                    return False
        else:
            logger.error("output.json file does not exist. No results available to download.")

            return False


if __name__ == '__main__':
    logger.info("**************************************")
    logger.info("API Sample tool downloadFiles starting...")
    logger.info("Logs can be found in the current directory, processing.log")
    downloadFiles()

File: ContentAnalyzer/C#/ContentAnalyzer/DownloadFiles.cs
/*disclaimer of warranties.
 this code is sample code created by ibm corporation. ibm grants you a
 nonexclusive copyright license to use this sample code example. this
 sample code is not part of any standard ibm product and is provided to you
 solely for the purpose of assisting you in the development of your
 applications. this example has not been thoroughly tested under all
 conditions. ibm, therefore cannot guarantee nor may you imply reliability,
 serviceability, or function of these programs. the code is provided "as is",
 without warranty of any kind. ibm shall not be liable for any damages
 arising out of your or any other parties use of the sample code, even if ibm
 has been advised of the possibility of such damages. if you do not agree with
 these terms, do not use the sample code.
 copyright ibm corp. 2019 all rights reserved.
 to run, see readme.md
 * */

using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Net;
using System.IO;
using RestSharp;
using Newtonsoft.Json;
using Newtonsoft.Json.Linq;
using NLog;

namespace ContentAnalyzer
{
    public class DownloadFiles
    {
        private static Logger logger = LogManager.GetCurrentClassLogger();

        /// <summary>
        /// Read the configuration.json file to get the request parameters required for the get request.
        /// will read the analyzer ID saved in the uploadFile function and use it to get the processing status first
        /// if the status is Completed, will download the output, if the status is inProgress, will wait for 5 second and check the status again
        /// the maximum loop to check the status is set to be 50
        /// </summary>
        /// <param></param>
        ///<returns>the response will be saved in the same output.json file, such as, "json":"downloaded","pdf":"downloaded","utf8":"downloaded"</returns>
        public bool DownloadFile()
        {

            ServicePointManager.ServerCertificateValidationCallback += (sender, certificate, chain, sslPolicyErrors) => true;
            ServicePointManager.SecurityProtocol = SecurityProtocolType.Tls12;

            DirectoryInfo currentDir = new DirectoryInfo(Directory.GetCurrentDirectory());
            string configFile = currentDir.Parent.Parent.FullName + @"/NLog.config";

            string functionalID = ContentAnalyzer.UploadFiles.configParams.functionalID;
            string password = ContentAnalyzer.UploadFiles.configParams.password;
            string LDAP_userName = ContentAnalyzer.UploadFiles.configParams.LDAP_userName;
            string LDAP_password = ContentAnalyzer.UploadFiles.configParams.LDAP_password;
            string mainURL = ContentAnalyzer.UploadFiles.configParams.mainURL.Trim();
            string apiKey = ContentAnalyzer.UploadFiles.configParams.apiKey;
            string outputDirectoryPath = ContentAnalyzer.UploadFiles.configParams.outputDirectoryPath;
            string[] types = getTypes(ContentAnalyzer.UploadFiles.configParams.responseType);
            string auth = "";
            if (functionalID != "" && password != "")
            {
                auth = ContentAnalyzer.UploadFiles.Base64Encode(functionalID + ":" + password);
            }
            else if (LDAP_userName != "" && LDAP_password != "")
            {
                auth = ContentAnalyzer.UploadFiles.Base64Encode(LDAP_userName + ":" + LDAP_password);
            } 

            string outputDir = @"../../../output.json";
            string outputFile = File.ReadAllText(@outputDir);
            if (outputFile != "")
            {
                JObject outputFiles = JObject.Parse(@outputFile);
                JArray uploadResults = JArray.Parse(outputFiles["upload_results"].ToString());

                var analyzers_unfinished = new List<JObject>();

                int no = 0;
                foreach (JObject uploadResult in uploadResults)
                {
                    JObject response = JObject.Parse(uploadResult["upload_response"].ToString());
                    JObject data = JObject.Parse(response["data"].ToString());
                    string id = data["analyzerId"].ToString();
                    JObject ID = new JObject(uploadResult);
                    ID["analyzerId"] = id;
                    ID["deleted"] = false;
                    ID["time_elapsed"] = "";
                    foreach(string type in types) {
                        ID[type] = "";
                    }
                    analyzers_unfinished.Add(ID);
                    no++;
                }

                if (Directory.Exists(outputDirectoryPath))
                {
                    foreach (string type in types)
                    {
                        if (! Directory.Exists(outputDirectoryPath + "/" + type))
                        {
                            DirectoryInfo path = Directory.CreateDirectory(outputDirectoryPath + "/" + type);
                        }
                    }
                }

                int maxLoop = 50;
                int loop = 1;
                var finished_analyzers = new List<JObject>();
                while (loop < maxLoop)
                {
                    logger.Info("Check processing status: loop " + loop);
                    System.Threading.Thread.Sleep(5000);
                    var latest_analyzers_unfinished = new List<JObject>(analyzers_unfinished);

                    foreach (JObject analyzer in latest_analyzers_unfinished)
                    {
                        string analyzerId = analyzer["analyzerId"].ToString();
                        logger.Info("Checking status of analyzerId: " + analyzerId);

                        string url = mainURL + "/contentAnalyzer/" + analyzerId;
                        var client = new RestClient(url);
                        var request = new RestRequest(Method.GET);

                        request.AddHeader("Authorization", "Basic " + auth);
                        request.AddHeader("apiKey", apiKey);

                        IRestResponse response = client.Execute(request);

                        if (response.StatusCode == System.Net.HttpStatusCode.OK)
                        {
                            string res = response.Content;
                            JObject resObj = JObject.Parse(res);
                            JObject dataObj = JObject.Parse(resObj["data"].ToString());
                            JArray statusDetails = JArray.Parse(dataObj["statusDetails"].ToString());
                            var final_analyzers = new List<JObject>();

                            foreach (JObject statusDetail in statusDetails)
                            {
                                string typeDownload = statusDetail["type"].ToString().ToLower();
                                string status = statusDetail["status"].ToString();
                                if (status == "Completed" && analyzer[typeDownload].ToString() != "downloaded" && analyzer[typeDownload].ToString() != "failed")
                                {
                                    string url_download = mainURL + "/contentAnalyzer/" + analyzerId + "/" + typeDownload;
                                    var client_download = new RestClient(url_download);
                                    var request_download = new RestRequest(Method.GET);

                                    request_download.AddHeader("Authorization", "Basic " + auth);
                                    request_download.AddHeader("apiKey", apiKey);

                                    IRestResponse response_download = client_download.Execute(request_download);

                                    logger.Info("Downloading the " + typeDownload + " output of the analyzerId: " + analyzerId);

                                    if (response_download.StatusCode == System.Net.HttpStatusCode.OK)
                                    {
                                        if (Directory.Exists(@outputDirectoryPath))
                                        {

                                            var res_download = response_download.Content;
                                            //string fileExt = Path.GetExtension(analyzer["filename"].ToString());
                                            string newFileName = "";
                                            if (typeDownload == "pdf")
                                            {
                                                newFileName = "New_" + Path.GetFileNameWithoutExtension(analyzer["filename"].ToString()) + ".pdf";
                                                //if (fileExt == ".pdf")
                                                //    newFileName = "New_" + Path.GetFileNameWithoutExtension(analyzer["filename"].ToString()) + ".pdf";
                                                //else
                                                //    newFileName = Path.GetFileNameWithoutExtension(analyzer["filename"].ToString()) + ".pdf";

                                                string outputPath = outputDirectoryPath + "/" + typeDownload + "/" + newFileName;
                                                System.IO.File.WriteAllBytes(@outputPath, response_download.RawBytes);
                                            }
                                            else if (typeDownload == "json")
                                            {
                                                newFileName = Path.GetFileNameWithoutExtension(analyzer["filename"].ToString()) + ".json";
                                                string outputPath = outputDirectoryPath + "/" + typeDownload + "/" + newFileName;
                                                JObject resJson = JObject.Parse(res_download);
                                                System.IO.File.WriteAllText(@outputPath, resJson["data"].ToString());
                                                analyzer["time_elapsed"] = statusDetail["timeElapsed"].ToString();

                                            }
                                            else
                                            {
                                                newFileName = Path.GetFileNameWithoutExtension(analyzer["filename"].ToString()) + ".txt";
                                                string outputPath = outputDirectoryPath + "/" + typeDownload + "/" + newFileName;
                                                System.IO.File.WriteAllText(@outputPath, res_download);
                                            }
                                            analyzer[typeDownload] = "downloaded";
                                        }
                                        else
                                        {
                                            logger.Error("Failed to download " + typeDownload + " output of the analyzerId: " + analyzerId);
                                            analyzer[typeDownload] = "failed";
                                        }
                                    }
                                    else {
                                        logger.Error("Output directory defined in config.json does not exist.");
                                    }

                                }
                                else if (status == "failed")
                                {
                                    analyzer[typeDownload] = "failed";
                                }
                            }

                            bool allCompleted = true;
                            foreach (string type in types)
                            {
                                if (analyzer[type].ToString() != "downloaded" && analyzer[type].ToString() != "failed")
                                {
                                    allCompleted = false;
                                }
                            }
                            if (allCompleted == true)
                            {
                                analyzers_unfinished.Remove(analyzer);
                                finished_analyzers.Add(analyzer);
                            }

                            final_analyzers.AddRange(finished_analyzers);
                            final_analyzers.AddRange(analyzers_unfinished);
                            //Update output
                            outputFiles["upload_results"] = new JArray(final_analyzers);
                            System.IO.File.WriteAllText(@outputDir, JsonConvert.SerializeObject(outputFiles));
                        }
                        else
                        {

                            logger.Error("Http error occurred when checking status of analyzerId: " + analyzerId);
                            foreach (string type in types)
                            {
                                analyzer[type] = "Failed";
                            }
                        }
                    }
                    logger.Info("Number of unfinished files: " + analyzers_unfinished.Count);
                    if (analyzers_unfinished.Count != 0)
                    {
                        loop++;
                    }
                    else
                    {
                        loop = maxLoop + 1;
                    }
                }
            }
            return true;
        }


        public static string[] getTypes(string options)
        {
            string[] types = options.Split(',').Select(p => p.Trim().ToLower()).ToArray();
            int i = 0;
            foreach(string type in types)
            {
                string type1 = type.Substring(1);
                types[i] = type1.Substring(0, type.Length - 2);
                i++;
            }

            return types;
        }

    }
}
